randomActivityGen.py

```diff
@@ -91,6 +92,88 @@ def setup_city_gates(net: sumolib.net.Net, stats: ET.ElementTree, gate_count: in
+        x, y = get_edge_pair_centroid(edge.getShape())
+        return get_perlin_noise(x, y, POPULATION_BASE)
+
+    # Pick out the one edge with highest perlin noise from each district and return these to later place school on
+    for district in districts:
+        district.sort(key=noise)
+        school_edges.append(district[-1])
+
+    return school_edges
+
+
+def setup_schools(net: sumolib.net.Net, stats: ET.ElementTree, school_count: int or None):
+    # Voodoo parameter, seems to be about the value for a couple of danish cities.
+    # In general one high school, per 5000-7000 inhabitant in a city
+    inhabitants_per_school = 5000
+
+    school_opening_earliest = 7 * 3600
+    school_opening_latest = 10 * 3600
+    school_closing_earliest = 13 * 3600
+    school_closing_latest = 17 * 3600
+    stepsize = int(0.25 * 3600)
+
+    # Creates a list of school start times, in seconds. Ranges from 7am to 10am, with 15min intervals
+    school_start_times = list(range(school_opening_earliest, school_opening_latest, stepsize))
+
+    # Creates a list of school end times, in seconds. Ranges from 13pm to 17pm, with 15min intervals
+    school_end_times = list(range(school_closing_earliest, school_closing_latest, stepsize))
+
+    xml_schools = stats.find('schools')
+    if xml_schools is None:
+        xml_schools = ET.SubElement(stats.getroot(), "schools")
+    if school_count is None:
+        # Calculate default number of schools, based on population if none input parameter
+        xml_general = stats.find('general')
+        inhabitants = xml_general.get('inhabitants')
+        num_schools_default = math.ceil(int(inhabitants) / inhabitants_per_school)
+
+        # Number of new schools to be placed
+        number_new_schools = num_schools_default - len(xml_schools.findall("school"))
+    else:
+        # Else place new number of schools as according to input
+        number_new_schools = school_count - len(xml_schools.findall("school"))
+
+    if number_new_schools == 0:
+        return
+    if number_new_schools < 0:
+        print(f"Warning: {school_count} schools was requested, but there are already {len(xml_schools)} defined")
+        return
+
+    # Find edges to place schools on
+    new_school_edges = find_school_edges(net, number_new_schools)
+
+    # Insert schools, with semi-random parameters
+    print("Inserting " + str(len(new_school_edges)) + " new schools")
+    for school in new_school_edges:
+        begin_age = random.choice(list(range(6, 19)))
+        end_age = random.choice(list(range(begin_age + 1, 26)))
+
+        ET.SubElement(xml_schools, "school", attrib={
+            "edge": str(school.getID()),
+            "pos": str(random.randint(0, 100)),
+            "beginAge": str(begin_age),
+            "endAge": str(end_age),
+            "capacity": str(random.randint(100, 500)),
+            "opening": str(random.choice(school_start_times)),
+            "closing": str(random.choice(school_end_times))
```

Comment:
This can be made more efficient by using `random.randint` and a bit of math

Reply:
I initially did that, but I thought this solution looks cleaner, and is easier to change the times, e.g "school_closing_latest = 17 * 3600" you can just change 17 to whatever hour you'd like.

Reply:
```suggestion
            "opening": str(random.randrange(school_open_earliest, school_open_latest, stepsize)),
            "closing": str(random.randrange(school_close_earliest, school_close_latest, stepsize))
```
The documentation says this is functionally equivalent to what you have written, see https://docs.python.org/3.8/library/random.html#random.randrange. It doesn't remove `school_closing_latest = 17 * 3600`, and it doesn't waste memory storing all possibilities

Reply:
Changed in 14c4b1ee5f472c9dff95c0699fe8d2dbb2a7dcfc